experiments/goofy/utils.py

```diff
@@ -0,0 +1,98 @@
+import functools
+import numpy as np
+import pandas as pd
+
+def make_bee(regime, x, y, *args, **kwargs):
+    """
+    Convenience function for turning a simple arg
+    based training function into a the bee format of argless
+    This can be used to pass state between swarm iterations
+    and share resources
+    Also optional to use
+    """
+    thestrkwargs = {key: str(value) for key, value in kwargs.items()}
+    return functools.partial(regime, x, y, *args, **kwargs), thestrkwargs
+
+def add_anindex(an_array:np.ndarray, anindex):
+    """
+    This is a crude way of adding a column onto a numpy array populated with a constant.
+    It just allows us to stay with numpy while we join multiple arrays
+    together while we reduce dimensions.
+    """
+    return np.c_[an_array, np.array([anindex for a in range(len(an_array))])]
+
+def flatten_sim(data_list, loss_list, xd, yd):
+    """
+    This is adds indices to lists of arrays to return single stacked
+    arrays using an new col in each array carrying the index. 
+    This is currently limited to simulations of equal domain
+    and target function (i.e. x and y is same). Could use a refactor to expand. 
+    """
+    xy = np.c_[xd, yd]
+    data_list = [add_anindex(data_list[a], a) for a in range(len(data_list))]
+    flat_data = data_list[0]
+    for i in range(1, len(data_list)):
+        flat_data = np.vstack((flat_data, data_list[i]))
+    loss_list = [add_anindex(loss_list[a], a) for a in range(len(loss_list))]
+    flat_loss = loss_list[0]
+    for i in range(1, len(loss_list)):
+        flat_loss = np.vstack((flat_loss, loss_list[i]))
+    
+    return flat_data, flat_loss, xy
+
+def unpacker(reslist:list, static_params:dict):
+    """
+    This takes the whole listed output from a hive experiment (i.e multiple swarms) 
+    and flattens it into four dataframes, with indices to note swarm.
+    Currently assumes that x and y are static parameters and don't change across 
+    the entire hive, but is intended to expand to allow that to adjust.
+    This function leaves data in numpy arrays and lists of dicts. No pandas required.
+    """
+    data_list = [x['results']['ypred'] for x in reslist]
+    loss_list = [x['results']['loss'] for x in reslist]
```

Comment:
Yes, of course that's true. But I'd defend the sensibility of this as a step on the development path for most mortals.  It works well for what was already envisioned.  And it's transparent enough to be added too manually to incorporate a different data object for an emerging requirement.  I appreciate that generalising the function fully so that it automatically can incorporate all other possible requirements without any alteration so that every future person can already use your code without any modification is ideal. But for non-genius programmers, its often necessary (and I would say, practical) to first get something working with explicitly named separate objects, then expand that set of objects manually for the next requirement, maybe two, then only take the ultimate generalisation step once you have a pretty good feel for there being a need for many other objects, and what kind of type and function those objects would have.  So that you generalise well, for real requirements.  And also so that don't create (in one step) a piece of code so powerful but artisan that a mere mortal looking at it doesn't understand how it already does everything they have and haven't thought of. And so doesn't use it.

Reply:
Yep, as I pointed out, limiting scope results in simpler code. And it was necessary given that you were trying to solve a tough problem (like writing a bunch of csv files, how do you tell R which files it needs to read in?).

My review was more that for it to move from `experiments` to `swarm` it needs to be held to a higher standard